src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs

```diff
@@ -0,0 +1,292 @@
+
+                    for (var i = 0; i < deferredTasks.Length; i++)
+                    {
+                        var task = deferredTasks[i];
+
+                        try
+                        {
+                            await task(scope);
+                        }
+                        catch (Exception e)
+                        {
+                            logger?.LogError(e,
+                                "Error while processing deferred task '{TaskName}' on tenant '{TenantName}'.",
+                                task.GetType().FullName, ShellContext.Settings.Name);
+                        }
+                    }
+
+                    await scope.BeforeDisposeAsync();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Terminate the shell, if released and in its last scope, by calling the related event handlers.
+        /// Returns true if the shell context should be disposed consequently to this scope being released.
+        /// </summary>
+        private async Task<bool> TerminateShellAsync()
+        {
+            // A disabled shell still in use is released by its last scope.
+            if (ShellContext.Settings.State == TenantState.Disabled)
+            {
+                if (Interlocked.CompareExchange(ref ShellContext._refCount, 1, 1) == 1)
+                {
+                    ShellContext.Release();
+                }
+            }
+
+            // If the context is still being released, it will be disposed if the ref counter is equal to 0.
+            // To prevent this while executing the terminating events, the ref counter is not decremented here.
+            if (ShellContext._released && Interlocked.CompareExchange(ref ShellContext._refCount, 1, 1) == 1)
+            {
+                var tenantEvents = _serviceScope.ServiceProvider.GetServices<IModularTenantEvents>();
+
+                foreach (var tenantEvent in tenantEvents)
+                {
+                    await tenantEvent.TerminatingAsync();
+                }
+
+                foreach (var tenantEvent in tenantEvents.Reverse())
+                {
+                    await tenantEvent.TerminatedAsync();
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        public async Task DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            var disposeShellContext = _disposeShellContext || await TerminateShellAsync();
+
+            _serviceScope.Dispose();
+
+            if (disposeShellContext)
+            {
+                ShellContext.Dispose();
+            }
+
+            // Decrement the counter at the very end of the scope
+            Interlocked.Decrement(ref ShellContext._refCount);
+
+            _disposed = true;
+        }
+
+        public void Dispose() => DisposeAsync().GetAwaiter().GetResult();
```

Comment:
Checked `_disposed` before calling `GetAwaiter().GetResult()` so it's not used in the general case

Reply:
For sure.